x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java

```diff
@@ -635,17 +635,33 @@ private void runRequestInterceptors(
             listener.onResponse(null);
         } else {
             final Iterator<RequestInterceptor> requestInterceptorIterator = requestInterceptors.iterator();
-            requestInterceptorIterator.next()
-                .intercept(requestInfo, authorizationEngine, authorizationInfo, new DelegatingActionListener<>(listener) {
-                    @Override
-                    public void onResponse(Void unused) {
-                        if (requestInterceptorIterator.hasNext()) {
-                            requestInterceptorIterator.next().intercept(requestInfo, authorizationEngine, authorizationInfo, this);
-                        } else {
-                            listener.onResponse(null);
+            while (requestInterceptorIterator.hasNext()) {
+                var res = requestInterceptorIterator.next().intercept(requestInfo, authorizationEngine, authorizationInfo);
+                if (res.isDone() == false) {
+                    res.addListener(new DelegatingActionListener<>(listener) {
+                        @Override
+                        public void onResponse(Void unused) {
+                            if (requestInterceptorIterator.hasNext()) {
+                                var nestedRest = requestInterceptorIterator.next()
+                                    .intercept(requestInfo, authorizationEngine, authorizationInfo);
+                                if (nestedRest.isDone() == false) {
+                                    nestedRest.addListener(this);
+                                }
+                            } else {
+                                listener.onResponse(null);
+                            }
                         }
-                    }
-                });
+                    });
+                    return;
+                }
+                try {
+                    res.rawResult();
+                } catch (Exception e) {
+                    listener.onFailure(e);
+                    return;
```

Comment:
I'd rather add something like this to `SubscribableListener` over exposing `rawResult()`:

```
    public boolean completeIfFailed(ActionListener<?> listener) {
        final Object currentState = state;
        if (currentState instanceof FailureResult result) {
            listener.onFailure(result.exception());
            return true;
        } else {
            return false;
        }
    }
```

Reply:
I went for just an `isSuccess` now like other solutions have. There isn't much of a reason to be tricky for the failure path and this way we have a fast solution for the hot success path. Let me know what you think :) Combining this approach with the get-results on listenable futures would allow neat simplifications + speedups in some places I believe (by avoiding allocating new listeners).